models/modules/loss.py

```diff
@@ -394,6 +398,362 @@ def compute_loss_G(self, netD, real, fake):
+
+        self.iters += 1
+
+    def plot_cutmix(self, real, fake, cutmix_img):
+        for i in range(real.size(0)):
+            windom_id = "cutmix_figure_batch_{}".format(i)
+            self.viz.close(win=windom_id)
+            combined = torch.cat((real[i], fake[i], cutmix_img[i]), 2)
+            combined = (combined - combined.min()) / (combined.max() - combined.min())
+
+            combined = combined.cpu().detach().numpy().transpose(1, 2, 0)
+            combined = (combined * 255).astype(np.uint8)
+            self.viz.image(
+                combined.transpose(2, 0, 1),
+                opts=dict(
+                    title="Batch {}".format(i + 1), caption="Real | Fake | Cutmix"
+                ),
+                win=windom_id,
+            )
+            time.sleep(5)
+
+    def compute_loss_D(self, netD, real, fake, fake_2):
+        """Calculate GAN loss for the discriminator
+        Parameters:
+            netD (network)      -- the discriminator D
+            real (tensor array) -- real images
+            fake (tensor array) -- images generated by a generator
+        Return the discriminator loss.
+        We also call loss_D.backward() to calculate the gradients.
+        """
+        super().compute_loss_D(netD, real, fake, fake_2)
+
+        # Real
+        pred_real_pixel, pred_real_bottleneck = netD(self.real)
+
+        loss_pred_real_pixel = self.criterionGAN(pred_real_pixel, True)
+        loss_pred_real_bottleneck = self.criterionGAN(pred_real_bottleneck, True)
+        self.loss_D_real = loss_pred_real_pixel + loss_pred_real_bottleneck
+
+        # Fake
+
+        lambda_loss = 0.5
+
+        def cutmix_real_fake_pairwise(real, fake):
+            assert (
+                self.real.size() == self.fake.size()
+            ), "Real and fake images should have the same dimensions."
+            masks = torch.stack(
+                [
+                    CutMix(self.real.size(2)).to(self.real.device)
+                    for _ in range(self.real.size(0))
+                ]
+            )
+            masks = masks.unsqueeze(1)
+            mixed_images = masks * fake + (1 - masks) * real
+            return mixed_images, masks
+
+        train_use_cutmix = True  # change this into True to use cutmix
+
+        if train_use_cutmix:
+            lambda_cutmix = 0.5
+            fake_input = self.fake.detach()
+            pred_fake_pixel, pred_fake_bottleneck = netD(fake_input)
+
+            cutmix_img, label_masks = cutmix_real_fake_pairwise(self.real, fake_input)
+            fake_cutmix_input = cutmix_img.detach()
+            pred_cutmix_fake_pixel, pred_cutmix_fake_bottleneck = netD(
+                fake_cutmix_input
+            )
+            cutmix_pixel_label = label_masks.repeat(1, 3, 1, 1)
+
+            consistent_pred_pixel = torch.mul(
+                pred_fake_pixel, 1 - cutmix_pixel_label
+            ) + torch.mul(pred_real_pixel, cutmix_pixel_label)
+            loss_cutmix_pixel = (
```

Comment:
@wr0124 are you sure this is the loss from the paper ?
It should be D(cutmix(inputs)) - cutmix(D(inputs)) no ?

Reply:
yes, you are right. in the paper, it is || D(cutmix(inputs)) - cutmix(D(inputs)) ||²  .  So, D(cutmix(inputs)) is pred_cutmix_fake_pixel in my code, and cutmix(D(inputs)) is consistent_pred_pixel in my code

Reply:
OK cool, you can probably use `MSELoss` instead of the norm and `**2`, that´d be more efficient probably.

Reply:
yes, you are right. The MSELoss is being used here.